hash/hash_comparator.go

```diff
@@ -2,54 +2,134 @@ package hash
 
 import (
 	"context"
+	"crypto/aes"
+	"crypto/cipher"
 	"crypto/subtle"
 	"encoding/base64"
+	"encoding/hex"
 	"fmt"
-	"regexp"
 	"strings"
 
 	"github.com/pkg/errors"
 	"golang.org/x/crypto/argon2"
 	"golang.org/x/crypto/bcrypt"
+	"golang.org/x/crypto/sha3"
 
 	"github.com/ory/kratos/driver/config"
 )
 
 var ErrUnknownHashAlgorithm = errors.New("unknown hash algorithm")
+var ErrUnknownHashFormat = errors.New("unknown hash format")
+var ErrNonceSizeIncorrect = errors.New("nonce size is incorrect")
 
-func Compare(ctx context.Context, password []byte, hash []byte) error {
-	if IsBcryptHash(hash) {
-		return CompareBcrypt(ctx, password, hash)
-	} else if IsArgon2idHash(hash) {
-		return CompareArgon2id(ctx, password, hash)
-	} else {
+func Compare(ctx context.Context, cfg *config.Config, password []byte, hash []byte) error {
+	algorithm, realHash, err := ParsePasswordHash(hash)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	switch algorithm {
+	case Argon2AlgorithmId:
+		return CompareArgon2id(ctx, cfg, password, realHash)
+	case BcryptAlgorithmId:
+		return CompareBcrypt(ctx, cfg, password, realHash)
+	case BcryptAESAlgorithmId:
+		return CompareBcryptAes(ctx, cfg, password, realHash)
+	default:
 		return ErrUnknownHashAlgorithm
 	}
 }
 
-func CompareBcrypt(_ context.Context, password []byte, hash []byte) error {
+func CompareBcrypt(_ context.Context, _ *config.Config, password []byte, hash string) error {
 	if err := validateBcryptPasswordLength(password); err != nil {
-		return err
+		return errors.WithStack(err)
 	}
 
-	err := bcrypt.CompareHashAndPassword(hash, password)
+	err := bcrypt.CompareHashAndPassword([]byte(hash), password)
 	if err != nil {
-		return err
+		return errors.WithStack(err)
 	}
 
 	return nil
 }
 
-func CompareArgon2id(_ context.Context, password []byte, hash []byte) error {
+func IsPasswordHash(hash []byte) bool {
+	_, _, err := ParsePasswordHash(hash)
+	return err == nil
+}
+
+func ParsePasswordHash(input []byte) (algorithm, hash string, err error) {
+	hashParts := strings.SplitN(string(input), "$", 3)
```

Comment:
Previously stored bcrypt hashes will not have the alg ID prefixed right? We have to handle those legacy hashes and replace them in the database somehow.

Reply:
No they will not. In this case I will add a migration if that is OK.